### DISCUSSION TITLE
SLURP Recipe with HuBERT

### DISCUSSION THREAD

**Discussion by mutiann** [speechbrain/speechbrain#1731] [Q&A]
Hi,

I'm trying to reproduce the results in the paper [A Fine-tuned Wav2vec 2.0/HuBERT Benchmark For Speech Emotion Recognition, Speaker Verification and Spoken Language Understanding](https://arxiv.org/abs/2111.02735) on SLURP benchmark using HuBERT. The paper presents some very good results of >89% IC accuracy using full fine-tuning of HuBERT-large, and mentions that the recipes are available at https://github.com/speechbrain/speechbrain/tree/develop/recipes, but I didn't manage to find the corresponding recipe, and numbers reported by the Direct (HuBERT) recipe in the SLURP directory are different. May I know if the recipe could be available?

**Comment by TParcollet** [comment]
@Moumeneb1 @BenoitWang we need your help here please.

**Comment by BenoitWang** [comment]
Hi @mutiann, thanks for your interest! Yes the recipe is available under https://github.com/speechbrain/speechbrain/tree/develop/recipes/SLURP/direct , to run the recipe just look at [here](https://github.com/speechbrain/speechbrain/tree/develop/recipes/SLURP#direct-recipe)

As far as the number, the IC accuracy reported in readme (87.54%) is with `PF-hbt-base`, you only need to change in yaml config:
1. `wav2vec2_hub: "facebook/hubert-base-ls960"` --> `wav2vec2_hub: "facebook/hubert-large-ll60k"`
2. `freeze_wav2vec2_conv: True` --> `freeze_wav2vec2_conv: False`

to be ready for an experiment with `EF-hbt-large` and for the best accuracy reported in the paper.

**Reply by mutiann** [reply]
Thank you! I've tried to train the model following the recipe, with the only exception that due to memory limitations I use batch_size=3 and accumulation factor=2, which shouldn't have any negative impact. I get the best dev IC accuracy of 90.4% at epoch 29, which gives test accuracy=88.9%. 
Do you have any suggestions to obtain a better reproduction?

**Reply by BenoitWang** [reply]
Hi @mutiann , first make sure that your evaluation is done with [this](https://github.com/pswietojanski/slurp/tree/master/scripts/evaluation), then try this [tokenizer](https://www.dropbox.com/s/w72vtnf273yjwu4/SLURP_51_unigram.model)(51 unigram) and change `output_neurons` to 51. And maybe try a bit more epochs.

**Reply by mutiann** [reply]
Thank you! It seems that the tokenizer is the same one as the one automatically fetched by SpeechBrain. May I know how many epochs did you use?

**Reply by BenoitWang** [reply]
If I make it right the tokenizer fetched automatically is unigram 58 https://huggingface.co/speechbrain/SLU-direct-SLURP-hubert-enc/tree/main. Maybe go for 10 more epochs to 45.

**Reply by mutiann** [reply]
Thanks! I just check the [tokenizer_58_unigram.model](https://huggingface.co/speechbrain/SLU-direct-SLURP-hubert-enc/blob/main/tokenizer_58_unigram.model) and it seems that it is actually a 51-token one. May I know what are the expected differences between them?

**Reply by BenoitWang** [reply]
Thanks for the check so this would be something to update in the huggingface repo... In this case just keep the original tokenizer and change `output_neurons` into 51 and it would be ok.

**Reply by mutiann** [reply]
I managed to get the model work and reached the accuracy. Thank you!
